ros_packages/voice_assistant/voice_assistant/chat.py

```diff
@@ -81,51 +83,104 @@ def chat(self, goal_handle: ServerGoalHandle):
             request: Chat.Goal = goal_handle.request
             chat_id: str = request.chat_id
             content: str = request.text
+            generate_code: bool = request.generate_code
 
             # get the personality that is associated with the request chat_id from the pib-api
             with self.voice_assistant_client_lock: 
                 successful, personality = voice_assistant_client.get_personality_from_chat(chat_id)
             if not successful:
                 self.get_logger().info(f"no personality found for id {chat_id}")
                 goal_handle.abort()
-                return Chat.Result()
+                return Chat.Result() # TODO
             
             # create the user message
             self.executor.create_task(self.create_chat_message, chat_id, content, True)
 
             # receive an iterable of tokens from the public-api
             description = personality.description if personality.description is not None else "Du bist pib, ein humanoider Roboter."
+            if generate_code: description = CODE_DESCRIPTION_PREFIX + description
+            print(description)
             with self.public_voice_client_lock:
                 tokens = public_voice_client.chat_completion(content, description)
 
-            curr_sentence: str = ""
-            prev_sentence: str | None = None
-            sentence_boundary = re.compile(r"[^\d | ^A-Z][\.|!|\?|:]")
+            # regex for indentifying sentences
+            sentence_pattern = re.compile(r"^(?!<pib-program>)(.*?)(([^\d | ^A-Z][\.|!|\?|:])|<pib-program>)", re.DOTALL)
+            # regex-pattern for indentifying visual-code blocks
+            code_visual_pattern = re.compile(r"^<pib-program>(.*?)</pib-program>", re.DOTALL)
```

Comment:
`sentence_pattern` and `code_visual_pattern` could be used as class variables and compiled upon Node initialization instead of recompiling them in every `chat()` call.

Reply:
These objects might end up being used in different threads in parallel. The documentation for python's re module does not mention anything about thread safety, so I think it is safer, to just compile them with each individual function call. I think the performance penalty here is neglectable, since the function is called rather infrequently and the overall runtime of the function will most of the times be several seconds long.